plugins/synapticsmst/synapticsmst-device.c

```diff
@@ -504,221 +532,833 @@ synapticsmst_device_get_flash_checksum (SynapticsMSTDevice *device,
+			for (i=0; i<16; i++){
+				if (readBuf[i] != tagData[i]){
+					match = FALSE;
+				}
+			}
+			if (match){
+				break;
+			}
+		}
+	}while(retryCount++ < MAX_RETRY_COUNTS );
+	if (rc && retryCount== MAX_RETRY_COUNTS) {
+		g_set_error_literal (error,	
+						G_IO_ERROR,	
+						G_IO_ERROR_INVALID_DATA,
+						"set tag vaild fail");
+		return FALSE;				
+	}
+	/* set tag invaild*/
+	rc = synapticsmst_common_rc_get_command (connection,
+						 UPDC_READ_FROM_EEPROM, 1, 
+						 (EEPROM_BANK_OFFSET * BankInUse + EEPROM_TAG_OFFSET + 15), 
+						 tagData);
+	if (rc) {
+		g_set_error_literal (error, 
+					G_IO_ERROR, 
+					G_IO_ERROR_INVALID_DATA,
+				    "Failed to read Tag from flash");
+		return FALSE;						
+	}
+		
+	do
+	{
+		if (tagData[0] != 0xFF)	{  
+		/* CRC8 is not 0xFF, erase last 4k of bank# */
+			eraseOffset = (EEPROM_BANK_OFFSET * BankInUse + EEPROM_BANK_OFFSET - 0x1000) / 0x1000; /* offset for last 4k of bank# */
+			rc = synapticsmst_device_set_flash_sector_erase(device, FLASH_SECTOR_ERASE_4K, eraseOffset);				
+		}
+		else{  /* CRC8 is 0xFF, set it to 0x00 */
+			tagData[1] = 0x00;
+			rc = synapticsmst_common_rc_set_command (connection,
+							 UPDC_WRITE_TO_EEPROM, 1, 
+							 (EEPROM_BANK_OFFSET * BankInUse + EEPROM_TAG_OFFSET + 15), 
+							 &tagData[1]);				
+		}
+
+		rc = synapticsmst_common_rc_get_command (connection,
+							UPDC_READ_FROM_EEPROM, 1, 
+							(EEPROM_BANK_OFFSET * BankInUse + EEPROM_TAG_OFFSET + 15), 
+							readBuf);
+		if (rc) {
+			g_set_error_literal (error, 
+						G_IO_ERROR, 
+						G_IO_ERROR_INVALID_DATA,
+						"Failed to read CRC from flash");
+		}
+		else{
+			if ((readBuf[0] == 0xFF && tagData[0] != 0xFF) || (readBuf[0] == 0x00 && tagData[0] == 0xFF))				{
+				break;
+			}
+		}
+	}while( retryCount++ < MAX_RETRY_COUNTS );
+	if (rc && retryCount== MAX_RETRY_COUNTS) {
+		g_set_error_literal (error,	
+						G_IO_ERROR,	
+						G_IO_ERROR_INVALID_DATA,
+						"set tag invaild fail");
+		return FALSE;				
+	}
+	return TRUE;
+
+}
+
+gboolean
+synapticsmst_device_check_firmware_content (SynapticsMSTDevice *device,
+					GBytes *fw,
+					guint8 chipType,
+				    GError **error)
+{
+	guint8 *payload_data;
+	guint32 payload_len, payload_len_max;
+	guint32 checksum = 0;
+	guint32 offset = 0;
+	guint32 code_size = 0;
 	SynapticsMSTDevicePrivate *priv = GET_PRIVATE (device);
 	g_autoptr(SynapticsMSTConnection) connection = NULL;
 
+	if (chipType == TRUE) {
+		payload_len_max = PAYLOAD_SIZE_512K;
+	}
+	else {
+		payload_len_max = PAYLOAD_SIZE_64K; 	
+	}
 	/* get firmware data and check size */
 	payload_data = g_bytes_get_data (fw, NULL);
 	payload_len = g_bytes_get_size (fw);
-	if (payload_len > 0x10000 || payload_len == 0) {
+	if (payload_len > payload_len_max || payload_len == 0) {
 		g_set_error_literal (error,
-				     G_IO_ERROR,
-				     G_IO_ERROR_INVALID_DATA,
-				     "invalid file size");
+					 G_IO_ERROR,
+					 G_IO_ERROR_INVALID_DATA,
+					 "invalid file size");
```

Comment:
I think it's good practice here to not use a literal error, but show what size it was finding.  It at least makes debugging easier in the future.